src/compiler/parser.ts

```diff
@@ -2717,6 +2745,31 @@ module ts {
             return expr;
         }
 
+        function parseBindingElementInitializer(inParameter: boolean): Expression {
+            // BindingElement[Yield,GeneratorParameter,Await,AsyncParameter] : 
+            //      [+GeneratorParameter] BindingPattern[?Yield,?Await,GeneratorParameter] Initializer[In]opt
+            //      [+AsyncParameter] BindingPattern[?Yield,?Await,AsyncParameter] Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter] BindingPattern[?Yield,?Await] Initializer[In,?Yield,?Await]opt
+            // SingleNameBinding[Yield,GeneratorParameter,Await,AsyncParameter] : 
+            //      [+GeneratorParameter] BindingIdentifier[Yield] Initializer[In]opt
+            //      [+AsyncParameter] BindingIdentifier[Await] Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter] BindingIdentifier[?Yield,?Await] Initializer[In,?Yield,?Await]opt
+            var explicitFlags = ParserContextFlags.None;
+            var ambientFlags = ParserContextFlags.None;
+            if (!inGeneratorParameterContext() && !inAsyncParameterContext()) {

```

Comment:
Yes, different checks. Also, don't you want to be checking that if GP/AP are **on**, then you turn Yield/Await **off**?

I find it very confusing that explicitFlags get set implicitly, and ambientFlags get set explicitly. Why set ambientFlags? If they're ambient, they should be automatic. This may be my fundamental issue with understanding setContextParameters